### PR TITLE
Add support for checking Echo notifications

### PR DESCRIPTION
**Pull request by mrsrec**
Echo is broken on ScratchWikiSkin2 "out-of-the-box" because it adds weird, broken elements into a dropdown.

Now we're filtering those and adding our own custom notifications button to the navbar. This will show if you have unread notifications, and if so, link you to Special:Notifications

In the future it would be nice to implement the pop-up instead of linking the user, and generating the link MW instead of hard-coding it, but I believe this is low-priority since it will fix itself.

### CODE REVIEW

**Comment by mrsrec** [comment]
One unrelated thing that I forgot I added here: this fixes duplicating the ID "feet".

File: ScratchWikiSkinTemplate.php [hunk 1]
```diff
@@ -103,6 +103,37 @@ public function execute() {
...
+			$swsDropdownItems = "";
+			foreach ($this->data['personal_urls'] as $key => $tab) { 
+					if (!array_key_exists('href', $tab)) {
+						if (!array_key_exists('class', $tab)) {
+							$tab['class'] = '';
+						}
+						
+						$tab['class'] .= 'no-link';
+					}
+					if(isset($tab["class"]) && gettype($tab["class"]) == "array" && $tab["class"][0] == "mw-echo-alert"){
```
Comment (jacob-g) [inline ScratchWikiSkinTemplate.php:117]: a comment here to explain what the significance of all this is would be good

File: ScratchWikiSkinTemplate.php [hunk 2]
```diff
@@ -103,6 +103,37 @@ public function execute() {
...
+						}
+						continue;
+					}
+					$swsDropdownItems .= $this->getSkin()->makeListItem($key, $tab);
+			}
+			// ONLY output this if Echo extension is turned on. Otherwise don't put it there.
+			if(ExtensionRegistry::getInstance()->isLoaded("Echo")){
+			?>
+			<li class="link right notifications">
+				<a href="/wiki/Special:Notifications<?php /* This probably should be localized and not hardcoded, but it works*/ ?>"><div></div></a>
```
Comment (jacob-g) [inline ScratchWikiSkinTemplate.php:133]: You can (and should) use [`$this->makeLink`](https://doc.wikimedia.org/mediawiki-core/master/php/classBaseTemplate.html#a571b83eb61a5ed0bb91522c2a33eabd7)

File: ScratchWikiSkinTemplate.php [hunk 3]
```diff
@@ -103,6 +103,37 @@ public function execute() {
...
+			foreach ($this->data['personal_urls'] as $key => $tab) { 
+					if (!array_key_exists('href', $tab)) {
+						if (!array_key_exists('class', $tab)) {
+							$tab['class'] = '';
+						}
+						
+						$tab['class'] .= 'no-link';
+					}
+					if(isset($tab["class"]) && gettype($tab["class"]) == "array" && $tab["class"][0] == "mw-echo-alert"){
+						$swsUnread = true;
```
Comment (jacob-g) [inline ScratchWikiSkinTemplate.php:118]: a better name for this variable would be good; also you should define it as false before the loop, otherwise you risk getting a "variable undefined" error further down

File: ScratchWikiSkinTemplate.php [hunk 4]
```diff
@@ -200,7 +219,7 @@ public function execute() {
 					<div id="feet" style="margin: 0<?=$darkPref ? '' : '; display: none'?>"><?=$line?></div>
 					</div>
 				</div>
-				<ul id="feet">
+				<ul>
```
Comment (jacob-g) [inline ScratchWikiSkinTemplate.php:222]: What's the purpose of the change here?

Comment (jacob-g) [reply ScratchWikiSkinTemplate.php:222]: A further note: this seems to break the display of the footer links:
image

This is how it's supposed to look:
image

**Review by jacob-g — CHANGES_REQUESTED** [review]
The overall functionality and logic seems solid and I have tested it on my local and it indeed works. However there are a few style things, as well as a regression and the potential for a PHP error that need to be fix, all of which I called out in my comments. Please fix those when you get the chance and we can get this in.

File: ScratchWikiSkinTemplate.php (hunk 1 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php:117]: There already is a comment

File: ScratchWikiSkinTemplate.php (hunk 2 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php]: Not sure if that works here because it's an arbitrary page and there's no data. But I was able to localize the page using the same way the other links are

File: ScratchWikiSkinTemplate.php (hunk 3 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php:118]: There is no risk of a "variable undefined" error. The place it's read from uses `isset()` and it's deliberately *not* defined as false because that would break the isset check.

File: ScratchWikiSkinTemplate.php (hunk 4 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php]: Hmm. I was trying to fix there being two elements with the ID "feet" which is invalid. But it seems that the CSS was looking for both of them. I will fix this by converting it into a class.

File: ScratchWikiSkinTemplate.php (hunk 1 above)
Comment (jacob-g) [reply ScratchWikiSkinTemplate.php:117]: A comment on the specific line I highlighted would be good to explain what the purpose of this if statement is

File: ScratchWikiSkinTemplate.php (hunk 3 above)
Comment (jacob-g) [reply ScratchWikiSkinTemplate.php:118]: Fair enough

File: ScratchWikiSkinTemplate.php [hunk 5]
```diff
@@ -197,10 +217,10 @@ public function execute() {
 $link = "<a href=\"$url\" target=\"_blank\">$text</a>";
 $line = wfMessage('scratchwikiskin-dark-theme-feedback')->rawParams( $link )->inLanguage( $wgLang )->escaped();
 ?>
-					<div id="feet" style="margin: 0<?=$darkPref ? '' : '; display: none'?>"><?=$line?></div>
+					<div class="feet" style="margin: 0<?=$darkPref ? '' : '; display: none'?>"><?=$line?></div>
```
Comment (jacob-g) [inline ScratchWikiSkinTemplate.php:220]: This still doesn't work, since the CSS looks for `#feet`, not `.feet` - please make sure to test your code before committing it

**Review by jacob-g — CHANGES_REQUESTED** [review]
Please address my follow-up comments - there is still a regression and a code readability request

File: ScratchWikiSkinTemplate.php (hunk 5 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php:220]: I'm sorry! I did change the CSS to `.feet`, but forgot to submit the updated CSS.

File: ScratchWikiSkinTemplate.php (hunk 1 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php:117]: A new comment has been added

File: ScratchWikiSkinTemplate.php (hunk 4 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php]: Fixed

File: ScratchWikiSkinTemplate.php [hunk 6]
```diff
@@ -103,6 +103,39 @@ public function execute() {
...
+						continue;
+					}
+					$swsDropdownItems .= $this->getSkin()->makeListItem($key, $tab);
+			}
+			// ONLY output this if Echo extension is turned on. Otherwise don't put it there.
+			if(ExtensionRegistry::getInstance()->isLoaded("Echo")){
+			?>
+			<li class="link right notifications">
+				<a href="<?php echo Title::newFromText("Special:Notifications")->getLocalURL();?>"><div></div></a>
+				<?php echo(wfMessage( 'portal-url' )->inContentLanguage()->text())?>
```
Comment (jacob-g) [inline ScratchWikiSkinTemplate.php:135]: What is this doing here? This causes stray text to appear on my machine:
image

**Review by jacob-g — CHANGES_REQUESTED** [review]
Looks like you're almost there, but there's another regression.

File: ScratchWikiSkinTemplate.php (hunk 6 above)
Comment (mrsrec) [reply ScratchWikiSkinTemplate.php]: Oops. I added that during testing and forgot to remove it.

**Review by jacob-g — COMMENTED** [review]
At this point I think all my concerns have been addressed and this works on my local. Once I get the chance I'll test a bit more thoroughly then hopefully we can get this launched soon.

A side note, at some point in the future it may be good only to show the notification icon for logged-in users, but that doesn't need to be done in this PR.

**Comment by Kenny2github** [comment]
Removed myself from review as I pushed directly. I realize that the connection of some of these changes to the original purpose of the PR is tenuous at best, so I can rebase them out and into another PR if desired, but I feel like they have been overdue and should be included. @jacob-g I think this is now ready for your review.

**Review by jacob-g — CHANGES_REQUESTED** [review]
@Kenny2github 
A look at the code itself seems ok other than the one issue I outlined. Please fix that and I can start functional testing.

File: ScratchWikiSkinTemplate.php [hunk 7]
```diff
@@ -98,39 +95,58 @@ public function execute() {
...
+$echoUnread = 0;
+foreach ($this->data['personal_urls'] as $key => $tab) {
+	// additional class for non-link items
+	if (!array_key_exists('href', $tab)) {
+		if (!array_key_exists('class', $tab)) $tab['class'] = [];
+		if (is_array($tab['class'])) array_push($tab['class'], 'no-link');
+		else $tab['class'] .= ' no-link';
+	}
+	// treat "You have a new Talk page message" as 1 notification...
+	if (is_array($tab['class']) && in_array('mw-echo-alert', $tab["class"])) {
```
[the remark this answers is not in the document]
Comment (Kenny2github) [reply ScratchWikiSkinTemplate.php]: Should be fixed. Did not test with warning reporting enabled before.

**Review by jacob-g — COMMENTED** [review]
It seems to work ok now on my local with the latest fix. I'll hammer on it some more as time permits and we can hopefully get this in early next week.

**Review by jacob-g — COMMENTED** [review]
Looking at the code, this all looks good. I'll do more functional testing as time permits and then we can get this merged.
